Feat/Add AMM pool price + fixes for type issues

mlguys opened the pull request:
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
1. Added price retrieval from AMM pool information for XRPL connector
2. Added support for RLUSD token pairs in the XRPL markets dictionary
3. Various type fixes for XRPL connector code base

This PR should resolve issue https://github.com/hummingbot/hummingbot/issues/7490
This PR should address the bounty https://github.com/hummingbot/hummingbot/issues/7494

**Tests performed by the developer**:
Tested with XRPL connector unit tests.
Tested the price retrieval methods using a test script (included in the comments).

**Tips for QA testing**:
Test the XRPL connector with both regular trading pairs and AMM pool interactions. Verify new RLUSD trading pairs work as expected.

rapcmia commented:
Test commit `f28f08fd5b88404bb1f30f13b9285b97f0857000`
- Added RLUSD on trustline ✅ 
- Successfully connected XRPL ✅ 
  ```
  wss_node_url: quicknode
  wss_second_node_url: quicknode
  wss_third_node_url: getblock
  ```
- Setup v1 and v2 using RLUSD pair, Run tests for 30mins and 5hrs bots then review logs
  - No `..error on last traded prices` found, issue fixed as per tests results ✅ 
  - Orders placed and cancelled okay on intervals 60s and 300s
  - Getting same error `amminfo failed: isn’t a valid URI: scheme isnt ws or wss` posted earlier
    ![image](https://github.com/user-attachments/assets/85c2fcdc-ed2c-4305-8b47-187804ac73f0)
     - At the same time there different instances of `failed to process trade fills`
     - Observed that this happened on few of my buy orders that is paired with sell that are partially filled
     - While this happened, no orders stuck on exchange ✅ 
 
---

Got partially filled again on sell order 1.2680632E-7/24.1947339 RLUSD has been filled at 0E+8 XRP
```
2025-03-20 13:46:31,206 - 275986 - hummingbot.connector.client_order_tracker - INFO - The SELL order hbot-1742449583853875-SRDXP630bfa87c113c872557bb5e0e45f670a1c070 amounting to 1.2680632E-7/24.1947339 RLUSD has been filled at 0E+8 XRP.
2025-03-20 13:46:31,217 - 275986 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1742449591.0, "order_id": "hbot-1742449583853875-SRDXP630bfa87c113c872557bb5e0e45f670a1c070", "trading_pair": "RLUSD-XRP", "trade_type": "TradeType.SELL", "order_type": "OrderType.LIMIT", "price": "0E+8", "amount": "1.2680632E-7", "trade_fee": {"percent": "0", "percent_token": "RLUSD", "flat_fees": []}, "exchange_trade_id": "4B97370762D06BDC1875B31734C0851EC68BF12B6CC418E507C267D5C1CA800B", "exchange_order_id": "84623457-94891986", "leverage": 1, "position": "NIL", "event_name": "OrderFilledEvent", "event_source": "xrpl"}
```
  - Sell order cancelled at refresh time of 10mins
  - Got another set of orders again and sell order got partially filled and cancelled ok
    ```
    2025-03-20 13:56:40,097 - 275986 - hummingbot.connector.client_order_tracker - INFO - The SELL order hbot-1742450192195747-SRDXP630bfccbea0aa872557bb5e0e45f670a1c070 amounting to 8.824061E-8/24.2781968 RLUSD has been filled at 0E+8 XRP.
    2025-03-20 13:56:40,112 - 275986 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1742450200.0, "order_id": "hbot-1742450192195747-SRDXP630bfccbea0aa872557bb5e0e45f670a1c070", "trading_pair": "RLUSD-XRP", "trade_type": "TradeType.SELL", "order_type": "OrderType.LIMIT", "price": "0E+8", "amount": "8.824061E-8", "trade_fee": {"percent": "0", "percent_token": "RLUSD", "flat_fees": []}, "exchange_trade_id": "9D9907D3381C2EB6B74346EDB29F00266429589C0B26F0B4901EECF3A7321839", "exchange_order_id": "84623461-94892143", "leverage": 1, "position": "NIL", "event_name": "OrderFilledEvent", "event_source": "xrpl"}
    ```

Review sqlite:
- Tradefill
  ![image](https://github.com/user-attachments/assets/7d444c43-3383-4262-8f07-a1c4a0e8c2bf)
    - Most of the sell orders recorded are partially filled (36) orders with 0 XRP price and a very low amount
    - I have only  one instance that the sell order was partially filled until completed. ✅ 
      ```
      (1742465293000, 'RLUSD-XRP', 'SELL', 'LIMIT', 1, 0, '6D2F44AC86ECB146AA94DF3FFC1D2224DED397708FFB81DB0868583A52AD2027', 'hbot-1742465277608781-SRDXP630c34fe7c371872557bb5e0e45f670a1c070', 'Partially filled')
      (1742465828000, 'RLUSD-XRP', 'SELL', 'LIMIT', 14275826, 401916, '4D710D6AA54FB1297409D1ED0A8984434850D15FAE097DD184018F19B64DBA92', 'hbot-1742465277608781-SRDXP630c34fe7c371872557bb5e0e45f670a1c070', 'Partially filled')
      (1742465848000, 'RLUSD-XRP', 'SELL', 'LIMIT', 10604939, 401915, 'B46CD8E0E37629C2A1DF9EBC5E9F4872651FB8DEF170C1F1BA40EE634F1F4C37', 'hbot-1742465277608781-SRDXP630c34fe7c371872557bb5e0e45f670a1c070', 'Partially filled')
      ```
- Executor
    - Found 1 instance of `timelimit` order that was completed ✅ 
      - This is related to the sell order that was completed `hbot-1742465277608781-SRDXP630c34fe7c371872557bb5e0e45f670a1c070`
      - Closed by a market order `hbot-1742465878105841-BRDXP630c373b29f2f872557bb5e0e45f670a1c070`
    - Found 1 instance of failed order
      - Connector failed to fetch order hbot-1742456200362986-BRDXP630c132dbfc96872557bb5e0e45f670a1c070
        ```
        2025-03-20 15:38:53,247 - 275986 - hummingbot.connector.exchange.xrpl.xrpl_exchange.XrplExchange - ERROR - Order status not found for order hbot-1742456200362986-BRDXP630c132dbfc96872557bb5e0e45f670a1c070 (84623506), tx history: [{'close_time_iso': '2025-03-20T07:36:42Z', 'hash': '286BC74C39473FA0F6ED983865718A883AD9188BAECDFD8A1A68D1C539C050EC', 'ledger_hash': '66C8001E3687923FD7F856537F7479AEBC0F5EB16476EB805E7B654761DB26CD', 'ledger_index': 94893681, 'meta': {'AffectedNodes': [{'ModifiedNode': {'FinalFields': {'Balance': {'currency': '524C555344000000000000000000000000000000', 'issuer': 'rrrrrrrrrrrrrrrrrrrrBZbvji', 'value': '78.8324948'}, 'Flags': 1114112, 'HighLimit': {'currency': '524C555344000000000000000000000000000000', 'issuer': 'rMxCKbEDwqr76QuheSUMdEGf4B9xJ8m5De', 'value': '0'}, 'HighNode': '74a', 'LowLimit': {'currency': '524C555344000000000000000000000000000000', 'issuer': 'rB3KUz3Q3rJx53EYPiV7g67N3NoRckwuUi', 'value': '1000000000'}, 'LowNode': '0'}, 'LedgerEntryType': 'RippleState', 'LedgerIndex': '501A4BD1493FCE7FCE776286C5809F1982C46E87EE322DEF8076C06812E45F99', 'PreviousFields': {'Balance': {'currency': '524C555344000000000000000000000000000000', 'issuer': 'rrrrrrrrrrrrrrrrrrrrBZbvji', 'value': '54.4162474'}}, 'PreviousTxnID': 'FA9C92377B25B07C5FF01E6F52AECD6841DF36902EEFDE99853FAF1E757551A7', 'PreviousTxnLgrSeq': 94893646}}, {'ModifiedNode': {'FinalFields': {'Account': 'rB3KUz3Q3rJx53EYPiV7g67N3NoRckwuUi', 'Balance': '7405259', 'Flags': 0, 'OwnerCount': 4, 'Sequence': 84623507}, 'LedgerEntryType': 'AccountRoot', 'LedgerIndex': '67116F5FDE8B17A65098407574E220A2385FAE2E2EEB8BBEEFAD1CD242EF48F8', 'PreviousFields': {'Balance': '17375735', 'Sequence': 84623506}, 'PreviousTxnID': '3E68A50BFB886791CC07C1D76A32DA9CC74328CCD56FDEBF931597EB1577A3DA', 'PreviousTxnLgrSeq': 94893646}}, {'ModifiedNode': {'FinalFields': {'Balance': {'currency': '524C555344000000000000000000000000000000', 'issuer': 'rrrrrrrrrrrrrrrrrrrrBZbvji', 'value': '878086.4535960939'}, 'Flags': 1114112, 'HighLimit': {'currency': '524C555344000000000000000000000000000000', 'issuer': 'rMxCKbEDwqr76QuheSUMdEGf4B9xJ8m5De', 'value': '0'}, 'HighNode': '347', 'LowLimit': {'currency': '524C555344000000000000000000000000000000', 'issuer': 'rfmdBKhtJw2J22rw1JxQcchQTM68qzE4N2', 'value': '100000000'}, 'LowNode': '0'}, 'LedgerEntryType': 'RippleState', 'LedgerIndex': '67D34953F6086D7D50C18CB2DEA2CD7294C0D76A8EA495C666A83CD702250DF8', 'PreviousFields': {'Balance': {'currency': '524C555344000000000000000000000000000000', 'issuer': 'rrrrrrrrrrrrrrrrrrrrBZbvji', 'value': '878110.8698434939'}}, 'PreviousTxnID': 'A40206E073461B0D0F2A53BE9497F151204EFDA7DE402CB0265686E000AAC7C5', 'PreviousTxnLgrSeq': 94893678}}, {'ModifiedNode': {'FinalFields': {'Account': 'rfmdBKhtJw2J22rw1JxQcchQTM68qzE4N2', 'Balance': '199500258692', 'Flags': 0, 'OwnerCount': 32, 'Sequence': 99831229}, 'LedgerEntryType': 'AccountRoot', 'LedgerIndex': '6BC9286C5146B76D0D140873B509D72581AABEB79037BF1D9849830FBF5A9FE6', 'PreviousFields': {'Balance': '199490288236'}, 'PreviousTxnID': 'EC2401944EA31F53FF75536DE9F84302CFEE04C07E9C09AD8C3E875044C40BCF', 'PreviousTxnLgrSeq': 94893680}}, {'ModifiedNode': {'FinalFields': {'Account': 'rfmdBKhtJw2J22rw1JxQcchQTM68qzE4N2', 'BookDirectory': 'CFEC8953D22B1B50F20F46ECBCD51990A26BDB71951ED8265A0E81F39DE4C200', 'BookNode': '0', 'Flags': 0, 'OwnerNode': '1', 'Sequence': 99831227, 'TakerGets': {'currency': '524C555344000000000000000000000000000000', 'issuer': 'rMxCKbEDwqr76QuheSUMdEGf4B9xJ8m5De', 'value': '9971.061320676468'}, 'TakerPays': '4071715794'}, 'LedgerEntryType': 'Offer', 'LedgerIndex': 'EFBF82AEB4A56DF43F48FECBE6B88C8D5327B527E43C8E00CB8A9B682EAFE373', 'PreviousFields': {'TakerGets': {'currency': '524C555344000000000000000000000000000000', 'issuer': 'rMxCKbEDwqr76QuheSUMdEGf4B9xJ8m5De', 'value': '9995.477568076468'}, 'TakerPays': '4081686250'}, 'PreviousTxnID': '967E4F31B507932E7FCBB15B73FBE4D23A0CBB34F546EC7C17DAA57B62AA1212', 'PreviousTxnLgrSeq': 94893680}}], 'TransactionIndex': 5, 'TransactionResult': 'tesSUCCESS'}, 'tx_json': {'Account': 'rB3KUz3Q3rJx53EYPiV7g67N3NoRckwuUi', 'Fee': '20', 'Flags': 65536, 'LastLedgerSequence': 94893699, 'Memos': [{'Memo': {'MemoData': '68626F742D313734323435363230303336323938362D425244585036333063313332646266633936383732353537626235653065343566363730613163303730'}}], 'Sequence': 84623506, 'SigningPubKey': 'ED266A5658DD3E29EB938D33D186939ECA9F3FEFB45CF30E424C8850B299A26117', 'SourceTag': 19089388, 'TakerGets': '9999991', 'TakerPays': {'currency': '524C555344000000000000000000000000000000', 'issuer': 'rMxCKbEDwqr76QuheSUMdEGf4B9xJ8m5De', 'value': '24.4162474'}, 'TransactionType': 'OfferCreate', 'TxnSignature': 'E134AD5C1519904E650BFBB9AD473A54203F04DDCC1CA0978A0E0DD860FEB972B38473704BB5C40EA7EF958A834D79413FF389E8D8B1EBADC454DEA436A77A00', 'date': 795771402, 'ledger_index': 94893681}, 'validated': True}]
        2025-03-20 15:38:53,248 - 275986 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1742456333.0, "order_id": "hbot-1742456200362986-BRDXP630c132dbfc96872557bb5e0e45f670a1c070", "order_type": "OrderType.LIMIT", "event_name": "MarketOrderFailureEvent", "event_source": "xrpl"}
        2025-03-20 15:38:53,249 - 275986 - hummingbot.connector.client_order_tracker - INFO - Order hbot-1742456200362986-BRDXP630c132dbfc96872557bb5e0e45f670a1c070 has failed. Order Update: OrderUpdate(trading_pair='RLUSD-XRP', update_timestamp=1742456333.2472754, new_state=<OrderState.FAILED: 6>, client_order_id='hbot-1742456200362986-BRDXP630c132dbfc96872557bb5e0e45f670a1c070', exchange_order_id='84623506-94893699', misc_updates=None)
        ```
          - Check txn hash and offer was successfully posted
              https://xrpscan.com/tx/286BC74C39473FA0F6ED983865718A883AD9188BAECDFD8A1A68D1C539C050EC
              - I dont find any order that has been active nor stuck on exchange

Run balance command:
![image](https://github.com/user-attachments/assets/cdfb1739-5f20-401d-8f45-f38bd9ffc2bb)
- i assume its not related to this PR but it has been like this even after relaunch the hummingbot client couple of times

---

Run check_prices script-controller 
- Set SOLO-XRP to observe ✅
- Set RLUSD-XRP ✅ 
  ![image](https://github.com/user-attachments/assets/57ad6de2-1abc-4cc5-9f5f-4d8d6bca5fe6)
   - Observed each data displayed for every 10s and compare with exchange
   - One thing noticed is that some intervals displayed all same value for the 30-40s 
   - its possible that this is related to the limitation of nodes used?
 
Steps to reproduce:
1. Setup v1 and v2 strategy and use RLUSD pair
2. Start and observe connector behavior when placing orders, cancelling and filled 
3. Review sqlite and logs 

Attach logs,config and sqlite fils
[03202025.zip](https://github.com/user-attachments/files/19370101/03202025.zip)

mlguys commented:
Hey @rapcmia @david-hummingbot, I've added fixes for:
- Improve trading performance, should track order creation and fills better
- Added check for invalid node URI which could make getting price for AMM pool failed

For:

>* Observed each data displayed for every 10s and compare with exchange
>* One thing noticed is that some intervals displayed all same value for the 30-40s
>* its possible that this is related to the limitation of nodes used?

This is expected since:
- Getting last traded price could be delayed if there is no new trades (which usually happen for pairs on XRPL DEX)
- Getting best prices could be delayed if one of the price source (order book or AMM pool) keep having the price for prolonged time

rapcmia commented:
Test commit `2ddbbdce82a585ac2e5fbe3f857cdcd44606dbeb`
- check_prices.py ✅
- v1 pure-mm 
  - Observed user stream listener cancelled while there was a active order 
    ![image](https://github.com/user-attachments/assets/1e2623b4-43c3-4f50-a923-108d5c1a1738)
    - After waiting for refresh time to trigger, got this behavior where the fetching the order return not found
      ![image](https://github.com/user-attachments/assets/95ab3c6f-72d1-4020-b438-301cd0d2e6c0)
      - This behavior was reproduce (2/2)
      - Orders are stuck on exchange then connector starts placing orders again
      - Had to manually cancel orders for `RLUSD/XRP` pair
- Stop command initiated and order on exchange was never recognized
  ![image](https://github.com/user-attachments/assets/b7301d94-3d2e-4af4-ba69-2f81c710e4bf)
   - Steps to reproduce:
      1. Setup XRPL on v1 with RLUSD/XRP
      2. Observe behavior while waiting for orders to be refreshed

- v2 market_making.pmm-simple
  - Running RLUSD/XRP on v2 strategy seem having less error
  - Observed partial filled dust orders still happen
  - Active bot running using pmm_simple controller config for monitoring `in progress`

Logs and config file: [04082025.zip](https://github.com/user-attachments/files/19651333/04082025.zip)

mlguys commented:
Hey @rapcmia, I've added fixes for:
- Fixed some cases where connector didn't process order creation correctly thus affecting trade performance
- Improved performance of user data stream
- Fixed rate oracle didn't get the price from last trade price for connector that doesn't have rate oracle implemented

History command should display correct numbers for RLUSD-XRP now:

![CleanShot 2025-04-14 at 20 36 30@2x](https://github.com/user-attachments/assets/acd04433-d6cb-4d2b-87ce-b183094673c0)

rapcmia commented:
Test 49f3403f9675abea95889ac20bfdbe508fe34914
- Review logs:
 - There are no sell dust orders that are filled ✅ 
 - Initiate stop command then confirmed no stuck orders on exchange
 - However since `2025-04-17 01:14:06,016` there where errors related to status 429 or `too many requests`
    ![image](https://github.com/user-attachments/assets/9cb37225-2e1f-49a2-825e-1a80f9bb4836)
- There are instances of websockets.exceptions.InvalidStatusCode: server rejected WebSocket connection: HTTP 429 ❌
  ```
  2025-04-17 01:14:06,016 - 3669846 - hummingbot.connector.exchange.xrpl.xrpl_exchange.XrplExchange - WARNING - Failed to update balances. Error: server rejected WebSocket connection: HTTP 429
  Traceback (most recent call last):
    File "/home/yawnyunehh/hummingbot/hummingbot/7506/hummingbot/connector/exchange_py_base.py", line 939, in _update_all_balances
      await self._update_balances()
    File "/home/yawnyunehh/hummingbot/hummingbot/7506/hummingbot/connector/exchange/xrpl/xrpl_exchange.py", line 1504, in _update_balances
      await self._client_health_check()
    File "/home/yawnyunehh/hummingbot/hummingbot/7506/hummingbot/connector/exchange/xrpl/xrpl_exchange.py", line 1846, in _client_health_check
      await self._xrpl_query_client.open()
    File "/home/yawnyunehh/miniconda3/envs/hummingbot/lib/python3.12/site-packages/xrpl/asyncio/clients/async_websocket_client.py", line 215, in open
      await self._do_open()
    File "/home/yawnyunehh/miniconda3/envs/hummingbot/lib/python3.12/site-packages/xrpl/asyncio/clients/websocket_base.py", line 93, in _do_open
      self._websocket = await websocket_client.connect(self.url)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/yawnyunehh/miniconda3/envs/hummingbot/lib/python3.12/site-packages/websockets/legacy/client.py", line 655, in __await_impl_timeout__
      return await self.__await_impl__()
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/yawnyunehh/miniconda3/envs/hummingbot/lib/python3.12/site-packages/websockets/legacy/client.py", line 662, in __await_impl__
      await protocol.handshake(
    File "/home/yawnyunehh/miniconda3/envs/hummingbot/lib/python3.12/site-packages/websockets/legacy/client.py", line 329, in handshake
      raise InvalidStatusCode(status_code, response_headers)
  websockets.exceptions.InvalidStatusCode: server rejected WebSocket connection: HTTP 429
  ```
  - Unexpected error fetching order book snapshots ❌
    ```
    2025-04-17 01:14:39,842 - 3669846 - hummingbot.connector.exchange.xrpl.xrpl_api_order_book_data_source.XRPLAPIOrderBookDataSource - ERROR - Error fetching order book snapshot:  - 
    2025-04-17 01:14:39,842 - 3669846 - hummingbot.connector.exchange.xrpl.xrpl_api_order_book_data_source.XRPLAPIOrderBookDataSource - ERROR - Unexpected error fetching order book snapshot for RLUSD-XRP.
    Traceback (most recent call last):
      File "/home/yawnyunehh/hummingbot/hummingbot/7506/hummingbot/connector/exchange/xrpl/xrpl_api_order_book_data_source.py", line 66, in _request_order_book_snapshot
        orderbook_asks_info, orderbook_bids_info = await safe_gather(orderbook_asks_task, orderbook_bids_task)
                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/yawnyunehh/hummingbot/hummingbot/7506/hummingbot/core/utils/async_utils.py", line 22, in safe_gather
        return await asyncio.gather(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/yawnyunehh/hummingbot/hummingbot/7506/hummingbot/connector/exchange/xrpl/xrpl_api_order_book_data_source.py", line 102, in fetch_order_book_side
        raise ValueError(exception_msg)
    ValueError: Error fetching order book snapshot:  - 
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "/home/yawnyunehh/hummingbot/hummingbot/7506/hummingbot/core/data_type/order_book_tracker_data_source.py", line 158, in _request_order_book_snapshots
        snapshot = await self._order_book_snapshot(trading_pair=trading_pair)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/yawnyunehh/hummingbot/hummingbot/7506/hummingbot/connector/exchange/xrpl/xrpl_api_order_book_data_source.py", line 138, in _order_book_snapshot
        snapshot: Dict[str, Any] = await self._request_order_book_snapshot(trading_pair)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/yawnyunehh/hummingbot/hummingbot/7506/hummingbot/connector/exchange/xrpl/xrpl_api_order_book_data_source.py", line 81, in _request_order_book_snapshot
        raise Exception(f"Error fetching order book snapshot for {trading_pair}: {e}")
    Exception: Error fetching order book snapshot for RLUSD-XRP: Error fetching order book snapshot:  - 
    ```
      - Observed there are still instances of orders created then cancelled on set executor refresh time while the error occurs

Logs, config and sqlite: [04172025a.zip](https://github.com/user-attachments/files/19787273/04172025a.zip)


---

Running another bot today to compare and will check result after couple of hours

rapcmia commented:
Test commit `c8a34ce24ddc2af195db25ffa21d2d2694bb89ed`
- After 2hrs of running bot, node got rate limited then reviewed bot performance 
- No orders stuck on exchange upon `stop` and `exit` command ✅ 
- Review logs :
- A completed filled order getting cancellation requests ❌ 
  - On the log file below using `hbot-1745195561933444-SRDXP6333f01885a9344b60d582b215c63d3579c8d`, observed that it was completed and recorded on sqlite however still executing cancellation event 
  ```
  2025-04-21 08:32:49,532 - 1070583 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT BUY order hbot-1745195561933214-BRDXP6333f018859de44b60d582b215c63d3579c8d for 7.4196666 RLUSD-XRP at 0.471719.
  2025-04-21 08:32:49,545 - 1070583 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1745195569.0, "type": "OrderType.LIMIT", "trading_pair": "RLUSD-XRP", "amount": "7.4196666", "price": "0.471719", "order_id": "hbot-1745195561933214-BRDXP6333f018859de44b60d582b215c63d3579c8d", "creation_timestamp": 1745195561.0, "exchange_order_id": "84626452-95595167", "leverage": 1, "position": "NIL", "event_name": "BuyOrderCreatedEvent", "event_source": "xrpl"}
  2025-04-21 08:32:54,334 - 1070583 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT SELL order hbot-1745195561933444-SRDXP6333f01885a9344b60d582b215c63d3579c8d for 7.4048420 RLUSD-XRP at 0.472663.
  2025-04-21 08:32:54,397 - 1070583 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1745195574.0, "type": "OrderType.LIMIT", "trading_pair": "RLUSD-XRP", "amount": "7.4048420", "price": "0.472663", "order_id": "hbot-1745195561933444-SRDXP6333f01885a9344b60d582b215c63d3579c8d", "creation_timestamp": 1745195561.0, "exchange_order_id": "84626453-95595168", "leverage": 1, "position": "NIL", "event_name": "SellOrderCreatedEvent", "event_source": "xrpl"}
  2025-04-21 08:33:06,062 - 1070583 - hummingbot.connector.client_order_tracker - INFO - The SELL order hbot-1745195561933444-SRDXP6333f01885a9344b60d582b215c63d3579c8d amounting to 7.404841801630719/7.4048420 RLUSD has been filled at 0.4726628999999999498132289716 XRP.
  2025-04-21 08:33:06,077 - 1070583 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1745195586.0, "order_id": "hbot-1745195561933444-SRDXP6333f01885a9344b60d582b215c63d3579c8d", "trading_pair": "RLUSD-XRP", "trade_type": "TradeType.SELL", "order_type": "OrderType.LIMIT", "price": "0.4726628999999999498132289716", "amount": "7.404841801630719", "trade_fee": {"percent": "0.0", "percent_token": "RLUSD", "flat_fees": []}, "exchange_trade_id": "33D587425696AC2335F0C57EDD466EF8ECFA27B4B3362D35E52015831A9850DF", "exchange_order_id": "84626453-95595168", "leverage": 1, "position": "NIL", "event_name": "OrderFilledEvent", "event_source": "xrpl"}
  2025-04-21 08:33:16,079 - 1070583 - hummingbot.connector.client_order_tracker - WARNING - The order fill updates did not arrive on time for hbot-1745195561933444-SRDXP6333f01885a9344b60d582b215c63d3579c8d. The complete update will be processed with incomplete information.
  2025-04-21 08:33:16,142 - 1070583 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1745195596.0, "order_id": "hbot-1745195561933444-SRDXP6333f01885a9344b60d582b215c63d3579c8d", "base_asset": "RLUSD", "quote_asset": "XRP", "base_asset_amount": "7.404841801630719", "quote_asset_amount": "3.499994", "order_type": "OrderType.LIMIT", "exchange_order_id": "84626453-95595168", "event_name": "SellOrderCompletedEvent", "event_source": "xrpl"}
  2025-04-21 08:33:16,142 - 1070583 - hummingbot.connector.client_order_tracker - INFO - SELL order hbot-1745195561933444-SRDXP6333f01885a9344b60d582b215c63d3579c8d completely filled.
  ...
  
  # last cancellation event before we initiate stop command
  2025-04-21 10:42:23,875 - 1070583 - hummingbot.strategy.script_strategy_base - INFO - (RLUSD-XRP) Canceling the limit order hbot-1745195561933444-SRDXP6333f01885a9344b60d582b215c63d3579c8d. [clock=2025-04-21 02:42:23+00:00]
  ```
    - Txn of filled order https://xrpscan.com/tx/1AB4977955472EBD3F70532358FBAA3AB2DF4D1663789339E6D2FC3CD3D7C3AE
     - Observed that some instances of these orders happened when `order fill updates did not arrived on time` occur causing delay order updates.
     - Some of the orders below got same behavior
       - `hbot-1745197667304095-SRDXP6333f7f05c2c644b60d582b215c63d3579c8d` ❗ 
       - `hbot-1745200077150955-SRDXP633400ea91b1b44b60d582b215c63d3579c8d` ❗ 
- Review sqlite data:
  ![image](https://github.com/user-attachments/assets/d56ab1fc-0852-4185-8a26-61ff7d76a569)
    - Total of 28 orders completed with 14 each buy and sell order ✅ 
    - Executor table:
        - Match log file’s trade data ✅ 
        - Match hummingbot status data ✅ 

Steps to reproduce:
1. Start v2 pmm_simple for XRPL using RLUSD/XRP
2. Run for couple of hours then review logs, sqlite and txn

logs, config and sqlite: [04212025a.zip](https://github.com/user-attachments/files/19829510/04212025a.zip)

cardosofede commented:
just ran again the test that were failing

nikspz commented:
Should be fixed in [9a86926](https://github.com/hummingbot/hummingbot/pull/7506/commits/9a86926c8ee18ac6e2dead0baa03d7963d2af685) 👍
![image](https://github.com/user-attachments/assets/117c94d1-3376-4960-bc2e-a2188e4c45b4)

rapcmia reviewed (approved):
- Run tests on linux (ubuntu24.04) on source build
- Run tests on v2 strategies: grid_strike, pmm_simple, xrpl_arb_sample and xrp_liquidity_example ok 
- Run bots for couple of hours and review
	- Orders opened, canceled and filled (even partial)
	- Orders placed and canceled has txhash added 
	- Filled and partial filled are recorded ok
	- Review every closeTypes on sqlite ok 
	- Open and close position successfully 
	- Out of range buffer starts and reset when market price change
	- Observed delay on initialize stop command which waits orders cancelled complete on XRPL
	- When min_profitability reached, OrderType.AMM_SWAP used for placing orders on XRPL 
		- However strategy suffers from conversion of XRP-RLUSD on binance (not related to PR)
- Reported issues to dev has been address
- Orders are monitored on internal XRPL app

---

Note: Tests use third party RPC nodes
- There will be instances of getting rate limited since we are using free tier (approx 5-8hrs of runtime)
- For default nodes, xplcluster,s1 and s2 ripple nodes getting rate limit approx 3-5 hours of runtime 
- We recommend to use a paid RPC node for better performance

mlguys commented on scripts/xrpl_liquidity_example.py:
```diff
@@ -0,0 +1,481 @@
+        )
+
+        # Check connector status
+        safe_ensure_future(self.check_connector_status())
```
Fixed, thanks!

mlguys commented on hummingbot/connector/exchange/xrpl/xrpl_constants.py:
```diff
@@ -470,4 +477,209 @@
         "base_issuer": "rpakCr61Q92abPXJnVboKENmpKssWyHpwu",
         "quote_issuer": "",
     },
+    "XRP-RLUSD": {
```
Fixed, thanks!

mlguys commented on hummingbot/client/performance.py:
```diff
@@ -309,8 +309,11 @@ async def _initialize_metrics(self,
         self.cur_price = await RateOracle.get_instance().stored_or_live_rate(trading_pair)
-        if self.cur_price is None:
-            self.cur_price = Decimal(str(trades[-1].price))
+        if self.cur_price is None or self.cur_price.is_nan() or self.cur_price == s_decimal_0:
```
revert back to original code, thanks!

mlguys commented:
@rapcmia added commits to resolve change requests
